infrastructure/README.md

```diff
@@ -1,26 +1,56 @@
-In short, ansible will run commands to assert state on a target machine.
-Asserting state means, for example, that a certain system user should exist,
-if not, ansible will create it. In this ansible is idempotent by design.
-
-Ansible is driven by the concept of a playbook. It ties an inventory
-file, which contains a set of hosts defined by groups, to 'roles'. Roles
-can be thought of as an applications. The roles are defined by the
-folders in the 'roles' folder. Ansible roles have a standard file format
-where variables are defined. In this project the two files of most
-note are "defaults/main.yml", which defines all variables to be used
-in a role, and "tasks/main.yml" which defines a series of tasks to execute.
-
-It's also notewory that 'group_vars' can be used to define environment
-specific variables. Ansible has a precedence ordering for variables,
-'group_vars' have high precedence, they override most other values,
-'defaults' have lowest.
-So you may see for example a default variable that is defined like 
-`bot_lobby_uri: "{{ lobby_uri }}"`
-That essentially means there is no default value for the variable
-and we expect the variable to be defined in group_vars.
+
+* Deployments are idempotent: https://shadow-soft.com/ansible-idempotency-configuration-drift/
+
+* Ansible consists of three main components:
+  * roles: think of these as applications
+  * inventory files: lists servers or hosts by group
+  * playbooks: binds host groups to roles
+
+* Per environment configuration is done through hostgroups, defined in inventory
+files. For example the production inventory file will have a hostgroup called
+'production' and we we can then define a 'group_vars/production.yml' file that
+has variable values that only apply to the 'production' hosts. These variable
+values will 'override' any variable values defined in 'defaults'.
+
+## Variables
+
+* Ansible allows variables to be defined in many places with differing
+levels of precendence. For simplicity,  in this project we constrain where
+variables are defined.
+
+### Role Defaults
+
+These variables will be in 'roles/[role_name]/defaults/main.yml'.
+Roles should define a default value for all variables used exclusively
+by that role.
+
+### Group vars (all.yml - shared variables)
+
+'groups_vars/all.yml' is a file that applies to all host groups.
+Variables that are shared between multiple roles are defined here.
+
+### Group vars (environment/hostgroup specific)
+
+For variables that are both shared and environment specific, we will
+define them in a group_vars file. For example: "database_password"
+may be defined in "groups_vars/production.yml" to have one value
+on production, and also defined in "group_vars/vagrant.yml" to
+have another value when deploying to a local vagrant server.
+
+The inventory files should all have the same structure and
+define hostgroups that line up with the "group_var" file names.
+
+
+### Notes on variables
+
+* do not use `vars` folder
```

Comment:
Maybe a stupid question, but what exactly is the vars folder, and why shouldn't it be used?

Reply:
Role vars are number 15 out of the 22 levels of precedence and places where variables can be defined: https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#ansible-variable-precedence

Why not use vars?
It's a similar problem to using an overly strong CSS selector, when trying to override it, it can be really difficult and artificial to find a selector with a stronger strength.

In our case we should not need vars, just use defaults; essentially favoring the weakest precedence to give the most flexibility for overrides. 

'vars' are probably best when re-using an existing role via some sort of import, even then I'm not sure if that's really a cleaner why to go compared to using a group_var maybe, can depend on the context. Regardless, the point is that we're limiting where you can define variables to 3 or 4 locations to simplify that list of 22 down to just a few.